assets/src/design-system/components/swatch/swatch.js

```diff
@@ -0,0 +1,171 @@
+ */
+import styled, { css } from 'styled-components';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { themeHelpers } from '../../theme';
+import {
+  getOpaqueColor,
+  presetHasGradient,
+  presetHasOpacity,
+} from '../../../edit-story/components/panels/design/preset/utils';
+import generatePatternStyles from '../../../edit-story/utils/generatePatternStyles';
+
+const RADIUS_LARGE = 32;
+const RADIUS_SMALL = 24;
+const ICON_SIZE = 32;
+
+const Transparent = styled.div`
+  width: 100%;
+  height: 100%;
+  position: absolute;
+  background-image: conic-gradient(
+    ${({ theme }) => theme.colors.fg.tertiary} 0.25turn,
+    transparent 0turn 0.5turn,
+    ${({ theme }) => theme.colors.fg.tertiary} 0turn 0.75turn,
+    transparent 0turn 1turn
+  );
+  background-size: 8px 8px;
+  border-radius: 100%;
+`;
+
+const SwatchButton = styled.button.attrs({ type: 'button' })`
+  cursor: pointer;
+  background-color: transparent;
+  border-color: transparent;
+  padding: 0;
+  border-width: 0;
+  display: block;
+  width: ${({ isSmall }) => (isSmall ? RADIUS_SMALL : RADIUS_LARGE)}px;
+  height: ${({ isSmall }) => (isSmall ? RADIUS_SMALL : RADIUS_LARGE)}px;
+  border-radius: 100%;
+  overflow: hidden;
+  position: relative;
+  ${({ disabled }) =>
+    disabled &&
+    css`
+      opacity: 0.4;
+      cursor: default;
+    `}
+  ${({ theme }) => themeHelpers.focusableOutlineCSS(theme.colors.border.focus)};
+
+  ::after {
+    content: '';
+    position: absolute;
+    left: 0;
+    top: 0;
+    width: 100%;
+    height: 100%;
+    border-radius: 50%;
+    border: 1px solid ${({ theme }) => theme.colors.divider.primary};
+    box-sizing: border-box;
+  }
+`;
+
+const presetCSS = css`
+  display: block;
+  width: 100%;
+  height: 100%;
+  font-size: 13px;
+  position: relative;
+
+  svg {
+    width: ${ICON_SIZE}px;
+    height: ${ICON_SIZE}px;
+    position: absolute;
+    top: calc(50% - ${ICON_SIZE / 2}px);
+    left: calc(50% - ${ICON_SIZE / 2}px);
+    filter: drop-shadow(0px 0px 4px rgba(0, 0, 0, 0.4));
+  }
+`;
+const SwatchItem = styled.div`
+  ${presetCSS}
+  ${({ pattern }) => generatePatternStyles(pattern)}
+  transform: rotate(${({ displaySplit }) => (displaySplit ? -45 : 0)}deg);
+
+  svg {
+    color: ${({ theme }) => theme.colors.fg.primary};
+    transform: rotate(${({ displaySplit }) => (displaySplit ? 45 : 0)}deg);
+  }
+`;
+
+const OpaqueColorWrapper = styled.div`
+  height: ${({ isSmall }) => (isSmall ? RADIUS_SMALL : RADIUS_LARGE)}px;
+  width: 50%;
+  overflow: hidden;
+  position: absolute;
+  left: 0;
+  top: 0;
+`;
+
+const OpaqueColor = styled.div`
+  height: ${({ isSmall }) => (isSmall ? RADIUS_SMALL : RADIUS_LARGE)}px;
+  width: ${({ isSmall }) => (isSmall ? RADIUS_SMALL : RADIUS_LARGE)}px;
+  position: absolute;
+  top: 0;
+  left: 0;
+  ${({ pattern }) => generatePatternStyles(pattern)}
+`;
+
+function Swatch({
+  pattern,
+  isDisabled = false,
+  isSmall = false,
+  children,
+  ...props
```

Comment:
Would `onClick` etc. come via `...props`? I wonder if for clarity, these main props should be written out for it being understandable which props should/can be used? E.g. `onClick` would probably be required, no? Or should we just use docs in this case? WDYT?

In the editor, in case of disabled swatches a tooltip is also shown (to explain why something is disabled). Should we add an optional tooltip text as a prop as well and display the tooltip in case it exists?

Reply:
I was considering renaming it to `SwatchButton` to make it clear, that this is just a fancy-looking button - to make it do anything, you need to attach `onClick`, etc. So yes, the idea was to use the extra props to attach actual functionality and labels.

Reply:
I also didn't want to include the tooltip, as you'd then also need the placement label - and then you just have a more compact version of the existing API with no actual advantage gained, IMO. I find it better to just wrap it in a tooltip if needed.